nixos/modules/hardware/device-tree.nix

```diff
@@ -38,14 +145,29 @@ in {
           '';
         };
 
+        filter = mkOption {
+          type = types.nullOr types.str;
+          default = null;
+          example = "*rpi*.dtb";
+          description = ''
+            Only include .dtb files matching glob expression.
+          '';
+        };
+
         overlays = mkOption {
           default = [];
-          example = literalExample
-            "[\"\${pkgs.device-tree_rpi.overlays}/w1-gpio.dtbo\"]";
-          type = types.listOf types.path;
+          example = literalExample ''
+            [
+              { name = "pps"; dtsFile = ./dts/pps.dts; }
+              { name = "spi";
+                dtsText = "...";
+              }
+              { name = "precompiled"; dtboFile = ./dtbos/example.dtbo; }
+            ]
+          '';
+          type = types.listOf overlayType;
```

Comment:
This is a breaking change like this. You can change the type to (untested)
```nix
types.listOf (types.coercedTo types.path (path: {
  name = baseNameOf path;
  dtboFile = path;
}) overlayType
```

for it to be backwards compatible

Reply:
I think this would be better as an `attrsOf` because it is much easier to override, although that is even more of a breaking change.

Reply:
Although I guess that makes establishing the order of the overlays more difficult.

Reply:
Hm true. I plan to make an RFC to deprecate `listOf` types (as good as possible) in favor of `attrsOf`. Ordering of values is one of the few things I haven't quite figured out yet.

Reply:
Looks good, thanks! I'll try to incorporate this and switch to `attrsOf` since the `name` is exactly `attr` name. Best to do it now rather than introducing another breaking change later.

Reply:
I think the order in which overlays are applied could matter - for example if I include `vendor.dtb`, which adds a device with some sensible defaults, and then `customization.dtb`, which overwrites some of these with my specific configuration.

Though I have to admit I'm not actually doing this right now, but it's a use-case I could imagine. IMO it would be better to
leave this as a list until the ordering issue mentioned above is figured out.

Reply:
I think the `name` attribute here is somewhat misleading, it doesn't specify anything about where the overlay is applied, or in what order. It's just an internal name given to the generated store path, so `attrsOf` is not the right type here.

Reply:
Good point, let's keep it as `listOf` for now with the added compatibility using `coercedTo`.